Add scalability story to ENSNode self-hosting module (#2146)
namehash/ensnode: docs/ensnode.io/src/content/docs/docs/self-host/scalability.mdx

---
title: ENSNode Scalability

> description: Understand how ENSNode's architecture enables horizontal scaling through PostgreSQL replication.

Reviewer: ```suggestion
description: Understand how ENSNode's architecture enables horizontal scaling.
```

sidebar:
  label: Scalability
  order: 3
---

import { LinkCard } from '@astrojs/starlight/components';

> ENSNode's architecture is designed for excellent scalability. By separating write and read workloads and building on PostgreSQL, you can scale your ENSNode deployment to handle high request volumes without impacting indexing performance.

Reviewer: ```suggestion
ENSNode's architecture achieves effectively limitless horizontal scalability, no matter how large your query demands might be. By separating read / write workloads and making use of PostgreSQL's async replication you can scale your ENSNode deployment to handle any request volumes while continuing to index the latest ENS data without impact to indexing performance.
```

> ## Single Writer, Many Readers

Reviewer: ```suggestion
## Horizontally scale ENSApi
```

The ENSNode architecture naturally separates writes from reads:

- **ENSIndexer** is the sole writer to ENSDb. It processes blockchain events and updates the database with the latest ENS state.
- **ENSApi** is a read-only service. It serves GraphQL and REST API requests by querying ENSDb, but never writes to it.

> This separation insulates ENSIndexer from high request volumes. Whether you serve 10 requests per second or 10,000, the indexing process remains unaffected because ENSApi does not compete with ENSIndexer for database write locks or indexer resources.

Author: The idea is for ENSApi instances to only use ENSDb read-replicas. Then, the ENSDb primary instance remains unaffected.

> This separation insulates ENSIndexer from high request volumes. Whether you serve 10 requests per second or 10,000, the indexing process remains unaffected because ENSApi does not compete with ENSIndexer for database write locks or indexer resources.

Reviewer: ```suggestion
The best first step for increasing your ENSNode's overall horizontal scalability is to horizontally scale ENSApi. You can go far with this, but if your request volumes become exceptionally high there is potential CPU/IO contention in a single "primary" ENSDb instance.
```

> ## Scale ENSDb with PostgreSQL Replication

Reviewer: ```suggestion
## Horizontally Scale ENSDb
```

ENSDb is built on PostgreSQL, which provides mature, production-proven mechanisms for horizontal scaling. The most common approach is **asynchronous replication**:

> 1. **Primary (master)**: ENSIndexer instance writes to a single ENSDb primary instance.

Reviewer: ```suggestion
1. **Primary ENSDb**: ENSIndexer writes to your single ENSDb primary instance. You can fully isolate this ENSDb primary instance from any incoming API requests coming through ENSApi.
```

> 2. **Replicas**: The primary ENSDb instance asynchronously replicates its data to one or more ENSDb read replicas.

Reviewer: ```suggestion
2. **Replica ENSDb**: Configure the asynchronous replication of your primary ENSDb instance to one or more ENSDb read replica instances.
```

> 3. **Read scaling**: You can connect any number of ENSApi instances to the ENSDb primary instance or to any ENSDb replica.

Reviewer: ```suggestion
3. **Read scaling**: Connect your horizontally scaled ENSApi instances exclusively to your ENSDb read replica instances if you wish to fully isolate extreme levels of API request volumes from indexing.
```

> Because ENSApi is read-only, it can safely connect to an ENSDb replica without risking stale writes or replication conflicts. As your query load grows, you simply add more PostgreSQL replicas for the ENSDb primary instance and more ENSApi instances pointing to them.

Reviewer: @tk-o Suggest to apply this feedback from Copilot

> ## Deployment Example

Reviewer: ```suggestion
## Going Massive
```

> A scalable ENSNode deployment might look like this:

Reviewer: ```suggestion
An ENSNode deployment with _massive_ horizontal scaling might look like this:
```

![ENSNode Scalable Deployment Architecture](/ensnode-scalability.svg)

- **One ENSIndexer instance** → writes to the ENSDb primary instance
- **One ENSDb primary instance** → replicates asynchronously to `M` replicas

> - **Multiple ENSApi instances** → each of `N` instances connected to the ENSDb primary instance or any ENSDb replica (often load-balanced across replicas)

Reviewer: This is the massive horizontal scalability case, right? Therefore, suggest to say that they only connect to replicas and no ENSApi instances connect to the primary.

- **One ENSRainbow instance** → co-located with ENSIndexer instance on the same local network for fast label healing

> This architecture lets you scale reads independently of writes, matching your infrastructure to your actual traffic patterns.

Reviewer: ```suggestion
Only the largest ENSNode deployments will need such a complex setup. Still, it's good to know that if you need massive scalability, we have you covered! The ENSNode architecture lets you scale reads independently of writes, matching your infrastructure to your actual traffic patterns.
```
